### PR TITLE
{AKS} Refactor a thirteenth part of the create sub-command in acs module

### PR DESCRIPTION
**Pull request by FumingZhang**
**Description**

* refactor identity & identity profile & auto upgrade profile

Unit test result:
- "covered_lines": 1045,
- "num_statements": 1116,
- "percent_covered": 93.63799283154121,
- "missing_lines": 71,
- "excluded_lines": 0

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/azure-cli/azure/cli/command_modules/acs/decorator.py [hunk 1]
```diff
@@ -380,6 +401,51 @@ def get_name(self) -> str:
...
+            location = self.mc.location
+            read_from_mc = True
+
+        # skip dynamic completion & validation if option read_only is specified
+        if read_only:
+            return location
+
+        # dynamic completion
+        if not read_from_mc and location is None:
+            location = _get_rg_location(
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/decorator.py:430]: what if failed to get the location?

File: src/azure-cli/azure/cli/command_modules/acs/decorator.py [hunk 2]
```diff
@@ -436,9 +502,19 @@ def get_ssh_key_value_and_no_ssh_key(self) -> Tuple[str, bool]:
...
-            no_ssh_key=no_ssh_key, ssh_key_value=ssh_key_value
-        )
+        if not no_ssh_key:
+            try:
+                if not ssh_key_value or not is_valid_ssh_rsa_public_key(
+                    ssh_key_value
+                ):
+                    raise ValueError()
+            except (TypeError, ValueError):
+                shortened_key = truncate_text(ssh_key_value)
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/decorator.py:512]: what's the behaviour of truncate_text?
to me even print one shortened_key is not acceptable.

File: src/azure-cli/azure/cli/command_modules/acs/decorator.py [hunk 3]
```diff
@@ -478,11 +553,16 @@ def _get_dns_name_prefix(
...
         if dynamic_completion:
-            dns_name_prefix = _get_default_dns_prefix(
-                name=self.get_name(),
-                resource_group_name=self.get_resource_group_name(),
-                subscription_id=self.get_intermediate("subscription_id"),
-            )
+            name = self.get_name()
+            resource_group_name = self.get_resource_group_name()
+            subscription_id = self.get_subscription_id()
+            # Use subscription id to provide uniqueness and prevent DNS name clashes
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/decorator.py:559]: original behaviour?

File: src/azure-cli/azure/cli/command_modules/acs/decorator.py [hunk 4]
```diff
@@ -603,22 +636,33 @@ def _get_vm_set_type(self, read_only: bool = False, **kwargs) -> Union[str, None
...
+        # the value verified by the validator may have case problems, and we will adjust it by default
         if not read_from_mc:
-            vm_set_type = _set_vm_set_type(
-                vm_set_type=vm_set_type,
-                kubernetes_version=self.get_kubernetes_version(),
-            )
+            kubernetes_version = self.get_kubernetes_version()
+            if not vm_set_type:
+                if kubernetes_version and StrictVersion(kubernetes_version) < StrictVersion("1.12.9"):
+                    print("Setting vm_set_type to availabilityset as it is \
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/decorator.py:644]: what's this print for?

File: src/azure-cli/azure/cli/command_modules/acs/decorator.py (hunk 1 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/decorator.py:430]: In function `_get_rg_location`, we actually call ResourceManagementClient to get the location "property" of the provided resource group (which is a required parameter). If the operation failed, it should be handled properly (raise an exception) by ResourceManagementClient.

File: src/azure-cli/azure/cli/command_modules/acs/decorator.py (hunk 3 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/decorator.py:559]: Yes, this is the original behavior defined in `_get_default_dns_prefix`.

File: src/azure-cli/azure/cli/command_modules/acs/decorator.py (hunk 4 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/decorator.py:644]: As a hint to user? This is the original behavior defined in `_set_vm_set_type`.

File: src/azure-cli/azure/cli/command_modules/acs/decorator.py (hunk 2 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/decorator.py:512]: Just truncate the string, by default the maximum length is 70 characters. Besides, this is the original behavior defined in `_validate_ssh_key`.

**Review by andyliuliming — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**